### PR TITLE
interfaces: add global gsettings interfaces

### PR DESCRIPTION
**Pull request by jdstrand**
When connected, applications can access the global gsettings of the user's
session. This interface is restricted because it gives privileged access to
sensitive information stored in gsettings (such as passwords) and allows
adjusting settings of other applications.

In the future gsettings will gain AppArmor integration to allow safe,
application-specific access to gsettings, but for now this interface can be
used to unblock people until that work is completed.

Closes: LP#1576308

### CODE REVIEW

**Comment by mvo5** [comment]
:+1:

File: snap/implicit_test.go [hunk 1]
```diff
@@ -56,7 +56,7 @@ func (s *InfoSnapYamlTestSuite) TestAddImplicitSlotsOnClassic(c *C) {
 	c.Assert(info.Slots["unity7"].Interface, Equals, "unity7")
 	c.Assert(info.Slots["unity7"].Name, Equals, "unity7")
 	c.Assert(info.Slots["unity7"].Snap, Equals, info)
-	c.Assert(info.Slots, HasLen, 17)
+	c.Assert(info.Slots, HasLen, 18)
```
Comment (zyga) [inline snap/implicit_test.go]: Can this be changed to say "more than 1" instead.

```c.Assert(len(info.Slots) > 1, Equals, true)```

Each new interface has to touch this.

File: overlord/ifacestate/ifacemgr_test.go [hunk 2]
```diff
@@ -451,7 +451,7 @@ func (s *interfaceManagerSuite) TestDoSetupProfilesAddsImplicitSlots(c *C) {
 	// Ensure that we have slots on the OS snap.
 	repo := mgr.Repository()
 	slots := repo.Slots(snapInfo.Name())
-	c.Assert(slots, HasLen, 17)
+	c.Assert(slots, HasLen, 18)
```
Comment (zyga) [inline overlord/ifacestate/ifacemgr_test.go]: Can this be changed to say "more than 1" instead.

c.Assert(len(info.Slots) > 1, Equals, true)

Each new interface has to touch this

File: interfaces/builtin/gsettings.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    interface="ca.desrt.dconf.Writer"
+    peer=(label=unconfined),
+`
+
+const gsettingsConnectedPlugSecComp = `
+# Description: Can access global gsettings of the user's session. Restricted
+# because this gives privileged access to sensitive information stored in
+# gsettings and allows adjusting settings of other applications.
+
+# dbus
```
Comment (zyga) [inline interfaces/builtin/gsettings.go]: Isn't there some abstraction that lets us say "uses dbus"?

File: interfaces/builtin/gsettings.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+# Description: Can access global gsettings of the user's session. Restricted
+# because this gives privileged access to sensitive information stored in
+# gsettings and allows adjusting settings of other applications.
+# Usage: reserved
+
+#include <abstractions/dbus-session-strict>
+
+#include <abstractions/dconf>
+owner /{,var/}run/user/*/dconf/user w,
+owner @{HOME}/.config/dconf/user w,
```
Comment (zyga) [inline interfaces/builtin/gsettings.go]: Isn't this irrelevant since $HOME is really snap-specific home anyway?

File: interfaces/builtin/gsettings.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+const gsettingsConnectedPlugAppArmor = `
+# Description: Can access global gsettings of the user's session. Restricted
+# because this gives privileged access to sensitive information stored in
+# gsettings and allows adjusting settings of other applications.
+# Usage: reserved
+
+#include <abstractions/dbus-session-strict>
+
+#include <abstractions/dconf>
+owner /{,var/}run/user/*/dconf/user w,
```
Comment (zyga) [inline interfaces/builtin/gsettings.go]: Is there a problem when writes use dconf and end up in the real ``/home/zyga/.config/dconf/user`` but reads are made from ``$HOME/.config/dconf/user`` (aka ``/home/zyga/snaps/$SNAP_NAME/$SNAP_REVISION/.config/dconf/user``?

File: interfaces/builtin/gsettings.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+`
+
+// NewGsettingsInterface returns a new "gsettings" interface.
+func NewGsettingsInterface() interfaces.Interface {
+	return &commonInterface{
+		name: "gsettings",
+		connectedPlugAppArmor: gsettingsConnectedPlugAppArmor,
+		connectedPlugSecComp:  gsettingsConnectedPlugSecComp,
+		reservedForOS:         true,
+		autoConnect:           false,
```
Comment (zyga) [inline interfaces/builtin/gsettings.go]: It should auto connect (snaps don't work by default without it) it should be reviewed by the store.

File: interfaces/builtin/gsettings.go (hunk 3 above)
Comment (jdstrand) [inline interfaces/builtin/gsettings.go]: This is seccomp, not apparmor so 'no'. I mentioned in other PRs that we probably do want to have a shorthand for certain things and dbus as a client is probably the most reasonable. That said, I think anything like that should be followed up in another PR.

File: interfaces/builtin/gsettings.go (hunk 4 above)
Comment (jdstrand) [inline interfaces/builtin/gsettings.go]: No, this is correct for 'global session gsettings' (that'll make more sense in my next response).

File: interfaces/builtin/gsettings.go (hunk 5 above)
Comment (jdstrand) [inline interfaces/builtin/gsettings.go]: Yes, that is absolutely a problem and not one which this PR is trying to address. It is important to remember this is for the 'global session gsettings' and as such, snaps need to use it as it is currently implemented, which is accessing /run/user/<uid>/dconf/user and /home/zyga/.config/dconf/user.

The path /home/zyga/snaps/$SNAP_NAME/$SNAP_REVISION/.config/dconf/user is unknown to gsettings as implemented and it cannot be used. Also, the future plan of apparmor gsettings mediation means that apps can continue to use the well-known paths (ie, not ~/snaps/...) and dconf will mediate access to snap-specific directories in the dconf database itself rather than using the actual filesystem.

File: interfaces/builtin/gsettings.go (hunk 6 above)
Comment (jdstrand) [inline interfaces/builtin/gsettings.go]: Well, this can be said of any interface that doesn't auto-connect, no? gsettings is used as a password store, contains other sensitive information and allows manipulating the user's session and other apps' settings, so I felt it prudent to not autoconnect.

**Comment by zyga** [comment]
Can you please merge this with master

File: interfaces/builtin/gsettings.go (hunk 6 above)
Comment (jdstrand) [inline interfaces/builtin/gsettings.go]: I'm going to treat this like 'home' and have it autoconnect and adjust store auto-approval accordingly.

File: interfaces/builtin/gsettings.go (hunk 5 above)
Comment (zyga) [inline interfaces/builtin/gsettings.go]: Ah, this makes sense, thanks for explaining it.

File: overlord/ifacestate/ifacemgr_test.go [hunk 7]
```diff
@@ -454,7 +454,7 @@ func (s *interfaceManagerSuite) TestDoSetupProfilesAddsImplicitSlots(c *C) {
 	// NOTE: This is not an exact test as it duplicates functionality elsewhere
 	// and is was a pain to update each time. This is correctly handled by the
 	// implicit slot tests in snap/implicit_test.go
-	c.Assert(len(slots) > 17, Equals, true)
+	c.Assert(len(slots) > 18, Equals, true)
```
Comment (zyga) [inline overlord/ifacestate/ifacemgr_test.go]: I think we can stop changing this line :-)
